ngraph/core/src/op/gru_cell.cpp

```diff
@@ -165,6 +165,102 @@ void op::v3::GRUCell::pre_validate_and_infer_types()
         ".");
 }
 
+void op::v3::GRUCell::validate_and_infer_types()
+{
+    std::vector<ngraph::PartialShape> input_param{};
+
+    auto merged_batch_size = Dimension::dynamic();
+    auto merged_hidden_size = Dimension::dynamic();
+    auto result_et = element::dynamic;
+
+    // Copy all inputs for further validation
+    for (size_t i = 0; i < get_input_size(); i++)
+    {
+        input_param.push_back(get_input_partial_shape(i));
+    }
+
+    // Get input partial shape for all inputs
+    const auto& x_pshape = get_input_partial_shape(0);
+    const auto& ht_pshape = get_input_partial_shape(1);
+    const auto& w_pshape = get_input_partial_shape(2);
+    const auto& r_pshape = get_input_partial_shape(3);
+    const auto& b_pshape = get_input_partial_shape(4);
+
+    validate_input_rank_dimension(input_param);
+
+    // Validate input types and save result for output type
+    NODE_VALIDATION_CHECK(
+        this,
+        element::Type::merge(result_et, result_et, get_input_element_type(0)) &&
+            element::Type::merge(result_et, result_et, get_input_element_type(1)) &&
+            element::Type::merge(result_et, result_et, get_input_element_type(2)) &&
+            element::Type::merge(result_et, result_et, get_input_element_type(3)) &&
+            element::Type::merge(result_et, result_et, get_input_element_type(4)),
+        "Element types for X, initial_hidden_state, W, R and B do not match.");
+
+    // Merge batch_size dimension across all inputs to evaluate output[0] dimension
+    NODE_VALIDATION_CHECK(this,
+                          Dimension::merge(merged_batch_size, merged_batch_size, ht_pshape[0]) &&
+                              Dimension::merge(merged_batch_size, merged_batch_size, x_pshape[0]),
+                          "Parameter batch_size not matched for ht_pshape and x_pshape.");
```

Comment:
General note which applies to all validation checks like this. Please strive not to use implementation variable names here. Rather than that use input names used in documentation. This message should also be digestible for a framework user.

Reply:
Corrected.